toolchain: nrfjprog.py fix and extension

siriobalmelli opened the pull request:
This is a patch series to fix a bug in nrfjprog.py and extend it to add a `sector_` target that flashes with `--sectorerase`.

All details in commit messages for each patch.

siriobalmelli committed:
nrfjprog.py: lint file using flake8

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
nrfjprog.py: fix typing bug on user input

On certain combinations of python+ZSH:

"nrfjprog.py", line 49, in select_devices
    if number.lower() == "q":
AttributeError: 'int' object has no attribute 'lower'

Fix by explicitly casting return value of input() to a string

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

siriobalmelli committed:
toolchain: add 'sector_' target to perform a sectorerase flash

The current 'flash_' target performs:
- 'nrfjprog --eraseall'
- 'nrfjprog --program'

This is essentially the same as 'nrfjprog --chiperase --program',
and flashes '${target}_merged.hex' onto the device, erasing everything.

With this new 'sector_' target, user can perform
an 'nrfjprog --sectorerase --program' which is useful
to push updates/fixes without re-provisioning.

After this change, the following targets are available:

- <target> : build <target>.hex
- sector_<target> : flash <target>.hex with --sectorerase
- merge_<target> : build <target>_merged.hex
- flash_<target> : flash <target>_merged.hex with --chiperase

Implement in such a way that the interface can later be
expanded to pass other nrfjprog options without breaking current usage.

Update relevant documentation.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicSemiconductor/nRF5-SDK-for-Mesh?pullRequest=7) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicSemiconductor/nRF5-SDK-for-Mesh?pullRequest=7) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicSemiconductor/nRF5-SDK-for-Mesh?pullRequest=7) it.</sub>

siriobalmelli committed:
nrfjprog.py: fix 'with' statement breakage on python2

Users which have a python2 stack (eg macOS) encounter an exception
on 'with'; which can be removed for identical functionality.

Signed-off-by: Sirio Balmelli <sirio@b-ad.ch>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/NordicSemiconductor/nRF5-SDK-for-Mesh?pullRequest=7) <br/>All committers have signed the CLA.

siriobalmelli commented:
Stale.
